fix(forms): ensure observable validators are properly canceled

kara opened the pull request:
…s are canceled

Observable subscriptions from previous validation runs should be canceled before a new subscription is created for the next validation run. Currently the subscription that sets the errors is canceled properly, but the source observable created by the validator is not. While this does not affect validation status or error setting, the source observables will incorrectly continue through the pipeline until they complete. This change ensures that the whole stream is canceled.

More validator changes to come to address larger issues with debouncing, observable chains, etc. Those changes will need to go through design review process.

DzmitryShylovich commented on packages/forms/src/validators.ts:
```diff
@@ -6,8 +6,12 @@
-import {toPromise} from 'rxjs/operator/toPromise';
+import {InjectionToken, ɵisObservable as isObservable, ɵisPromise as isPromise, ɵmerge as merge} from '@angular/core';
+import {Observable} from 'rxjs/Observable';
+import {forkJoin} from 'rxjs/observable/forkJoin';
```
need to add `forkJoin` and `map` [here](https://github.com/angular/angular/blob/master/packages/forms/rollup.config.js#L20)

kara replied:
Good catch!

kara committed:
fix(forms): ensure observable validators are properly canceled

Observable subscriptions from previous validation runs should be canceled
before a new subscription is created for the next validation run.
Currently the subscription that sets the errors is canceled properly,
but the source observable created by the validator is not. While this
does not affect validation status or error setting, the source
observables will incorrectly continue through the pipeline until they
complete. This change ensures that the whole stream is canceled.

AsyncValidatorFn previously had an "any" return type, but now it more
explicitly requires a Promise or Observable return type. We don't
anticipate this causing problems given that any other return type
would have caused a runtime error already.

IgorMinar reviewed (approved):
lgtm for api changes

kara commented:
In person LGTM from @robwormald
